cpp/src/arrow/adapters/avro/adapter.cc

```diff
@@ -0,0 +1,325 @@
+    avro_value_t child;
+    ArrayBuilder *child_builder;
+    Status result;
+
+    avro_value_get_size(&val, &actual_size);
+    static_cast<ListBuilder *>(builder)->Append(true);
+    child_builder = static_cast<ListBuilder *>(builder)->value_builder();
+    for (i = 0; i < actual_size; i++) {
+      avro_value_get_by_index(&val, i, &child, &map_key);
+      result = generic_read(child, child_builder);
+    }
+    return result;
+  };
+
+  Status read_map(const avro_value_t val, ArrayBuilder *builder) {
+    size_t num_values, i;
+    auto listB = static_cast<ListBuilder *>(builder);
+    auto structB = static_cast<StructBuilder *>(listB->child(0));
+    auto keyBuilder = static_cast<StringBuilder *>(structB->child(0));
+    auto valueBuilder = structB->child(1);
+    avro_value_t child;
+    const char *map_key;
+
+    avro_value_get_size(&val, &num_values);
+
+    listB->Append(true);
+    for (i = 0; i < num_values; i++) {
+      structB->Append(true);
+      avro_value_get_by_index(&val, i, &child, &map_key);
+      keyBuilder->Append(map_key);
+      generic_read(child, valueBuilder);
+    }
+    return Status::OK();
+  }
+
+  Status read_record(const avro_value_t val, ArrayBuilder *builder) {
+    avro_value_t child;
+    ArrayBuilder *child_builder;
+    StructBuilder *typed_builder;
+    Status result;
+
+    int container_length = builder->num_children();
+    typed_builder = static_cast<StructBuilder *>(builder);
+    typed_builder->Append();
+
+    for (auto i = 0; i < container_length; i++) {
+      avro_value_get_by_index(&val, i, &child, NULL);
+      child_builder = typed_builder->child(i);
+      result = generic_read(child, child_builder);
+    }
+    return result;
+  }
+
+};
+
+// Public API
+
+AvroArrowReader::AvroArrowReader() {
+  impl_.reset(new AvroArrowReaderImpl());
+  pool_ = default_memory_pool();
+}
+
+Status AvroArrowReader::ReadFromFileName(std::string filename, std::shared_ptr<Array> *out) {
+  avro_file_reader_t reader = NULL;
+  avro_file_reader(filename.c_str(), &reader);
+  return this->ReadFromAvroFile(&reader, out);
+}
+
+Status AvroArrowReader::ReadFromAvroFile(avro_file_reader_t *file_reader,
+                                         std::shared_ptr<Array> *out) {
+
+  std::unique_ptr<ArrayBuilder> builder;
+  std::shared_ptr<Array> res;
+  std::shared_ptr<DataType> datatype;
+  Status status;
+  int rval;
+  avro_value_iface_t *iface;
+  avro_value_t record;
+  ARROW_CHECK_OK(impl_->GenerateAvroSchema(file_reader, &datatype));
+  ARROW_CHECK_OK(MakeBuilder(this->pool_, datatype, &builder));
+
+  auto wschema = avro_file_reader_get_writer_schema(*file_reader);
+  iface = avro_generic_class_from_schema(wschema);
+  avro_generic_value_new(iface, &record);
+
+  while (true) {
+    rval = avro_file_reader_read_value(*file_reader, &record);
+    if (rval != 0)
+      break;
+    // decompose record into Python types
+    status = impl_->generic_read(record, std::move(builder).get());
```

Comment:
Instead of assigning the result to `status` you should call `RETURN_NOT_OK(impl_->...)` which will cause the function to return if a non `Status::OK()` value was returned when reading.